### ISSUE TITLE
[QST] How to best monitor RAM

### ISSUE THREAD

**Issue by Data-drone** [NVIDIA/spark-rapids#946]
**What is your question?**

What is the best way to monitor RAM consumption on GPU since that seems to be one of the main areas we need to monitor?

**Comment by revans2** [comment]
Sadly we don't have a great way to do it right now.  

Normally you could look at the GPU itself but because we use a pooling memory manager (RMM) the amount used is hidden because we have allocated a large block of memory up front.

For some operators we have a peak memory metric, but it just tells you about a single task not the actual state of the GPU and it is mostly looking at the inputs and output it does not tell how much memory is being used by the algorithm itself.

We could have visibility into exactly how much memory is being requested at any point in time by way of intercepting calls to the pooling memory manager, but have no easy way in Spark to export that data. We really just use it right now to see if we should try and free up memory to allow an allocation to succeed. Even then we do not know how fragmented the memory is.

@jlowe @tgravescs did I miss anything?

**Comment by jlowe** [comment]
Agree with @revans2 there isn't a great way to do it today.

We're already hooked into most GPU allocations via a custom RMM memory handler used for JVM callbacks on OOM, so tracking total allocated RMM memory from the pool would be very straightforward (albeit with no information on pool fragmentation).  However reporting this via Spark metrics in a useful way is unclear.  This seems to be ideally an executor metric reporting peak GPU memory, but I believe Spark's executor metrics are a fixed set and not extensible to cover concepts like GPU memory.  We may want to propose an Apache Spark feature to make executor metrics extensible by plugins, but that is probably not a simple, straightforward task.

That leaves us with task/SQL metrics.  We could track the maximum RMM pool allocation level, getting and resetting this max level atomically when a task completes.  Then at least one task will show the true maximum RMM pool memory used. However this gets a bit less useful when using the RAPIDS shuffle manager which caches shuffle outputs in GPU memory, as some of the GPU memory used and "charged" to a particular task/stage could have been produced in an earlier stage or even by an unrelated job.  In that case it may be useful for a task SQL metric to report both peak total RMM pool usage and peak spillable GPU memory usage, although even with those two metrics one can't assume both peaks were achieved at the same time.

**Comment by abellina** [comment]
On executor metrics, we can add metrics via the metrics registry in the executor. The registry is available via the `SparkEnv`. These are reported using the standard codahale metrics system, so it's not something that shows up in the UI, but can be piped to CSV at least with a Console sink, Prometheus, or other sinks as well.

I am working on a change to add some metrics for shuffle, but not for GPU memory metrics. If @jlowe/@revans2 think this would be a good idea perhaps we can add pool memory usage to these metrics.

**Comment by jlowe** [comment]
> These are reported using the standard codahale metrics system, so it's not something that shows up in the UI

Yes, sorry, I was referring to the UI.  We can definitely report custom executor metrics, but if they are not available in the standard Spark web UI and eventlog then IMHO the utility in practice is significantly diminished.  I agree that having _some_ way to get the metrics beats nothing. :smile:
